### PR TITLE
bpo-17422:language reference should specify restrictions on class namespace

### PR DESCRIPTION
**Pull request by ananthan-123**
https://bugs.python.org/issue17422

### CODE REVIEW

File: Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst]: Please attribute the patch to its author here.

File: Doc/reference/datamodel.rst
Comment (brandtbucher) [inline Doc/reference/datamodel.rst]: Please revert this unrelated whitespace change.

File: Doc/reference/datamodel.rst
Comment (brandtbucher) [inline Doc/reference/datamodel.rst]: No leading space here. Also, no need to add "instance":

```suggestion
is initialised as an empty :func:`dict`.
```

Separately, I have no idea if this is guaranteed by the language to be a `dict`. Before, it wasn't, but with this new change it certainly is.

If that's the case, then I'm against changing this line.

File: Doc/reference/datamodel.rst
Comment (brandtbucher) [inline Doc/reference/datamodel.rst:1949]: We lost a line with valuable info here. Please keep it and add your changes after.

File: Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst]: Let's make it clear that this was fixed:

```suggestion
The language reference now specifies restrictions on class namespaces.
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks @ananthan-123. No need to request reviews from 3 triagers and 4 core developers though, that's bordering on spam (especially if they have nothing to do with the issue/patch). One request should be fine, maybe a second if the first person doesn't respond after a week or two.

A few things I spotted here (I'm assuming that the wording in Eric's original patch from 7 years ago is correct):

**Review by brandtbucher — APPROVED** [review]
Looks good now. @ericsnowcurrently, care to review the content for this? It's your issue.

File: Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Documentation/2020-02-19-11-13-47.bpo-17422.g7_9zz.rst]: Since this isn't entirely your work @ananthan-123, let's credit Ethan here:

```suggestion
The language reference now specifies restrictions on class namespaces.  Adapted from a patch by Ethan Furman.
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Sorry, one more thing...

**Review by brandtbucher — APPROVED** [review]

**Review by gvanrossum — APPROVED** [review]

File: Doc/reference/datamodel.rst
Comment (ericsnowcurrently) [inline Doc/reference/datamodel.rst:1952]: Per PEP 520, the default class namespace must be an ordered mapping and the original text here was purposefully phrased.  While we've talked about requiring "dict" to be ordered ([in the language reference](https://docs.python.org/3/reference/datamodel.html#index-30)), that hasn't happened.  So changing the language spec text here to "dict" is inappropriate without a PEP.  Note that [elsewhere in the language reference](https://docs.python.org/3/reference/datamodel.html#creating-the-class-object) the "ordered mapping" phrasing still exists.

Comment (gvanrossum) [reply Doc/reference/datamodel.rst:1952]: Um, `dict` *is* required to be ordered. Since 3.7. See the 3.7. [what's new](https://docs.python.org/3/whatsnew/3.7.html) doc "the insertion-order preservation nature of dict objects has been declared to be an official part of the Python language spec."

If after that has sunk in you still think this PR is wrong, please submit a new PR to fix it! (Sorry for landing prematurely.)

Comment (cdonovick) [reply Doc/reference/datamodel.rst:1952]: Per my comment on the bpo I think @ericsnowcurrently is right but for the wrong reasons.   I have opened a PR correcting the wording.  https://github.com/python/cpython/pull/18682

**Pull request merged**
